src/passes/Monomorphize.cpp

```diff
@@ -247,12 +383,48 @@ struct CallContext {
   }
 
   // Checks whether an expression can be moved into the context.
-  bool canBeMovedIntoContext(Expression* curr) {
-    // Constant numbers, funcs, strings, etc. can all be copied, so it is ok to
-    // add them to the context.
-    // TODO: Allow global.get as well, and anything else that is purely
-    //       copyable.
-    return Properties::isSingleConstantExpression(curr);
+  bool canBeMovedIntoContext(Expression* curr,
+                             const ShallowEffectAnalyzer& effects) {
+    // Pretty much everything can be moved into the context if we can copy it
+    // between functions, such as constants, globals, etc. The things we cannot
+    // copy are now checked for.
+    if (effects.branchesOut || effects.hasExternalBreakTargets()) {
+      // This branches or returns. We can't move control flow between functions.
+      return false;
+    }
+    if (effects.accessesLocal()) {
+      // Reads/writes to local state cannot be moved around.
+      return false;
+    }
+    if (effects.calls) {
+      // We can in principle move calls, but for simplicity we avoid such
+      // situations (which might involve recursion etc.).
+      return false;
+    }
+    if (Properties::isControlFlowStructure(curr)) {
+      // We can in principle move entire control flow structures with their
+      // children, but for simplicity stop when we see one rather than look
+      // inside to see if we could transfer all its contents.
+      return false;
+    }
```

Comment:
But we already scan all its children, so the code would already handle moving entire control flow structures correctly, right?

Reply:
The problem would be that in general we can move a parent without moving the children. `(i32.eqz (local.get $anything))` can be partially optimized, with the `eqz` moved but not the local.get. But
```wat
(block $b
  (br $b)
)
```
will break if we move the `block` but not the `br`. So if we want to move the `block` we'd need to look ahead (really behind since we traverse in reverse) to see that all the children work out. That might be worth doing but I'm not sure, and it adds complexity.

Reply:
I see, and trying to partially move an if-else would also cause problems now that I think about it more.